### PR TITLE
Steps towards having a single beacon_node binary able to connect to all testnets

### PR DESCRIPTION
**Pull request by zah**
To test it out, just run

```bash
make beacon_node
build/beacon_node --network=altona
```

You may need to wipe out any previous data folder stored in `~/.cache/nimbus/BeaconNode` (adjusted for your OS)

### CODE REVIEW

File: beacon_chain/network_metadata.nim [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  stew/byteutils, nimcrypto/hash,
+  eth/common/[eth_types, eth_types_json_serialization]
+
+# ATTENTION! This file will produce a large C file, because we are inlining
+# genesis states as C literals in the generated code (and blobs in the final
+# binary). It makes sense to keep the file small and separated from the rest
+# of the module in order go gain maximum efficiency in incremental compilation.
+#
+# TODO:
+# We can compress the embedded states with snappy before embedding them here.
```
Comment (arnetheduck) [inline beacon_chain/network_metadata.nim:12]: hm, ideally we wouldn't do that at compile time, but rather move to a compressed version in the repo

**Comment by arnetheduck** [comment]
altona has a different fork version - that's not captured here, right? it'll break deposits

**Comment by arnetheduck** [comment]
to fix the fork issue, genesis fork version needs to be turned into a runtime variable that gets passed around, all the way to here: https://github.com/status-im/nim-beacon-chain/blob/devel/beacon_chain/spec/helpers.nim#L159

I get the feeling this pr might be addressing the issue from the wrong end - before we go here, it would make sense to be able to load metadata from the eth2 testnets repo and verify that the data is compatible with the binary by cross-checking the variables in the constants file with whatever the binary was compiled with (array sizes etc). to do this, we'd have a whitelist of constants that we don't compile into the binary - the rest would have to match - this would basically mean that in the current version, fork version wouldn't be on the whitelist, so I could try launching `beacon_node --metadata=shared/altona` and it would fail unless the binary conformed to the constants in that repo.

**Comment by arnetheduck** [comment]
once we have the runtime version working, adding a few "well-known" networks at compile time will be trivial, and it could use the same machinery to make sure that only compatible metadata is added to a particular binary - then if the user tries `--network=altona` and the binary is a `minimal` compile, it would print a message saying something like "metadata for altona not included, this binary was compiled with config options x y z that differ"

File: beacon_chain/network_metadata.nim (hunk 1 above)
Comment (zah) [reply beacon_chain/network_metadata.nim:12]: I see the current code as a stepping stone until we define a general mechanism for distributing recent snapshots (for clients who have been offline for longer than the weak subjectivity period). 

It does make sense to switch to a compressed version of the state now though. Perhaps it's not clear from the comment, but indeed the idea is to track the compressed version directly in the eth2-testnets repo.

**Comment by zah** [comment]
One of my goals was to make it very easy to distribute nbc. You just copy a single binary to a server and you are ready to go. Requiring the eth2-tesnets repo at run-time seemed off-putting.

Excellent point about the fork version. I'll make one final review of the entire Altona preset diff (when compared to mainnet) to make sure that no other details are missed.

**Comment by arnetheduck** [comment]
> You just copy a single binary to a server and you are ready to go.

sure - but without said checks (at runtime or compiletime), we're going to get users that use the wrong binary anyway - so my suggestion was merely to add the logic for comparing constants first which can be used to filter out non-conforming networks - this check can then be applied either at runtime or at compile time (ie I'm not suggesting that altona is not included at compile time - I'm suggesting that altona or any other testnet is only included _iff_ it is compatible)

File: beacon_chain/spec/signatures.nim [hunk 2]
```diff
@@ -114,23 +115,23 @@ func verify_attestation_signature*(
     blsFastAggregateVerify(pubkeys, signing_root.data, signature)
 
 # https://github.com/ethereum/eth2.0-specs/blob/v0.12.1/specs/phase0/beacon-chain.md#deposits
-func get_deposit_signature*(
-    deposit: DepositData,
-    privkey: ValidatorPrivKey): ValidatorSig =
-
+func get_deposit_signature*(preset: RuntimePreset,
```
Comment (arnetheduck) [inline beacon_chain/spec/signatures.nim:118]: :+1: for passing in the preset through arguments

**Comment by arnetheduck** [comment]
thinking about this branch a bit, it would make sense to store the runtime info + genesis info in the database as a record, so that given a database, one can resume operating without any further options - ie the database would be the single source of truth for startup and if any command-line parameters are incompatible with that database, the process would stop.

basically, the dependencies for a beacon chain are the parameters and the genesis, the genesis alone is not enough - also, in several places in the protocol, the genesis fork features (as opposed to `state.fork) - so we need access to this chain metadata information forever, but given that it's dynamic, storing it in the database seems better than storing it in the binary. 

At that point, the information in the binary (`--altona`) is used to bootstrap the database, but then the database takes over as source of truth - this would allow launching custom networks more easily because you create it, with parameters and all, once - and then you can just restart and keep it running, freely

**Comment by zah** [comment]
I had a different idea regarding data dirs. I think most users should never have to specify a custom `--data-dir` directory (eventually Nimbus will write to standard system locations such as `/var/cache`). To achieve this we can store the testnet-specific files in a directory nested within the data dir. To come up with a deterministic name for the directory, we can consider several inputs such as the `ForkVersion`, the genesis state root hash, the deposit contract address, etc. All of the required inputs should be present in a "network metadata" file, so it becomes a source of truth required only for custom networks. Admittedly, some complications in this approach are present in the "phase transition" that happens on genesis and the potential forks of a network.

**Pull request merged**